### ISSUE TITLE
Wrong export with timewarp on Linux

### ISSUE THREAD

**Issue by think-nice-things** [mltframework/shotcut#1177]
Hi,

If I split a clip into several pieces and arrange the pieces with differend speeds, the export is buggy on Linux.
(with the very same mlt file and same input files the export works correctly on Windows).

Let's have for example a clip split into three parts A, B, C and then arranged the parts A, B, C sequentially.
Then I change the speed for example to something like 1.4:A, 2:B and 0.9:C, let the clips still being sequentially arranged.

Buggy means - the export  sometimes _is_ correct, somtimes one part is missing and somtimes even a different part of the clip appears in the exported video.
So in the example above the resulting export may contain a white image instead of 1.4:A only, then (correctly) 2:B and 0.9:C.
Or it might contain even a different portion of the clip instead of 2:B -- and so on.

This is Shotcut-Version 21.10.31 on both windows and Linux, but the bug was already present in previous versions like 21.09.22.

**Comment by ddennedy** [comment]
I do not reproduce it using AppImage, Snap, or portable tar. I followed your steps, and I tested exporting 10 times. They all played the same and correctly. It is possible that the clip you are using does not seek well enough for speed change, and you need to Properties > Convert it. Properties > Speed and the Time Remap filter are more demanding on the reliability of seeking within the file; and that cannot be guaranteed for all video files, formats, codecs, and codec configurations. Properties > Convert transcodes to a known edit-friendly format. The test clip I used was not in one of these edit-friendly formats. It was simply a H.264+AAC Matroska file created with some unknown tool.

When you next reproduce it, you can view the export job log (in the job context menu), save it, and upload here.

From what I know about the code, I fail to see how there would be a difference between Windows and Linux... unless you were using a Linux build not made by us. Where did you get the Linux package and in what format?

**Comment by bmatherly** [comment]
> From what I know about the code, I fail to see how there would be a difference between Windows and Linux

Maybe a locale difference between the machines is exposing a problem on one machine and not the other?

Can you also share you localization settings on each computer?

**Comment by think-nice-things** [comment]
Hi,

it's indeed the locale, I can't believe.

## It exports **w/o errors**, if I set the locale to...

```bash
# LANG=C LANGUAGE=C LC_ALL=en_US.UTF-8 locale
LANG=C
LANGUAGE=C
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=en_US.UTF-8
# LANG=C LANGUAGE=C LC_ALL=en_US.UTF-8 ./shotcut
```

## It exports **with errors**, if I use my default locale settings:

```bash
# locale
LANG=de_DE.UTF-8
LANGUAGE=de_DE.UTF-8
LC_CTYPE="de_DE.UTF-8"
LC_NUMERIC="de_DE.UTF-8"
LC_TIME="de_DE.UTF-8"
LC_COLLATE="de_DE.UTF-8"
LC_MONETARY="de_DE.UTF-8"
LC_MESSAGES="de_DE.UTF-8"
LC_PAPER="de_DE.UTF-8"
LC_NAME="de_DE.UTF-8"
LC_ADDRESS="de_DE.UTF-8"
LC_TELEPHONE="de_DE.UTF-8"
LC_MEASUREMENT="de_DE.UTF-8"
LC_IDENTIFICATION="de_DE.UTF-8"
LC_ALL=de_DE.UTF-8
```

### errors

the errors are a bit more specifically:

* if the speed is below 1 (e.g. 0.8) then just a white (empty) sequence appears in the export
* if the speed is above 1 then a clip appears in the export but a wrong position of the source clip is used

### export log

attached is the export log of the bad export with german locale.
Note that my clip files do _not_ contain any special non-iso-latin characters.

[export_log.txt](https://github.com/mltframework/shotcut/files/7490827/export_log.txt)

**Comment by think-nice-things** [comment]
to get a correct export it's in fact sufficent to have

```bash
LC_NUMERIC=en_US.UTF-8
```
apparently the eport has difficulties with the (german) decimal separator ",".

**Comment by bmatherly** [comment]
Thanks for the extra testing.

I did some of my own tests using my Ubuntu computer.

My test steps are:
1) Open a file in Shotcut
2) In the properties, set the speed to "0.5" or "0,5" depending on the language
3) Export
4) Observe if the exported file is slowed down

### Test 1:
For my first test, did performed the steps above with the systems settings set for "English" language and formats.

In the UI, I entered "0.5" since that is what the UI requires. Everything worked as expected

### Test 2:
For my second test, I set the system settings for "German" language and formats:
![image](https://user-images.githubusercontent.com/821968/140663355-ae7fd983-7ff9-4d0c-8894-f0545ac4f25c.png)
In the UI, I entered "0,5" since that is what the UI requires. Everything worked as expected

## Test 3:
For my third test, I set the system settings back to "English" lanuage and formats and tried to specify the locale on the command line using:

LC_ALL=de_DE.UTF-8 LC_NUMERIC=de_DE.UTF-8 ./shotcut

In the UI, I entered "0,5" since that is what the UI requires. Everything worked as expected when previewing in the UI. **However, the export file only played at 1.0x speed, not 0.5x speed.** So maybe this reproduces your problem?

**Comment by ddennedy** [comment]
I tried test 3 and reproduced it, but when I set ONLY LC_NUMERIC=de_DE.UTF-8 (UI does show/require comma for decimal), I do NOT reproduce it. `melt`, which is used for export, is locale sensitive, but Shotcut write a locale independent XML (C) and runs `melt` with LC_NUMERIC=C only on Linux and macOS. It does not set LC_ALL and that probably makes the difference. On Windows, it sets LC_ALL instead due to how we had to inherit a `getlocale()` since it is not part of the platform or MinGW, and the implementation we inherited responds to that.

**Comment by ddennedy** [comment]
I did not reproduce Test 2 either.
Now, for test 3, the root cause is rather complicated. The problem is that the timewarp producer is using `atof()` directly in its constructor, and that function is locale-sensitive. MLT reads the numeric locale specified in XML (`LC_NUMERIC="C` as written by Shotcut). Then, all mlt_properties functions use other conversion functions that take a locale argument. Then, for legacy reasons, the locale is not passed to the service (plugin) constructors but set immediately after conversion. However, due to the way timewarp works, it needs to do some heavy lifting in its constructor and unable to use a specific locale in its mlt_properties at a later point. For reasons such as this and our inability to control what happens in dependencies resulting in bugs, Shotcut moved away from relying upon this locale override in XML. However, I made it only set MLT's LC_NUMERIC environment variable since I thought functions would be looking at that specifically if set. [I was wrong](https://www.gnu.org/software/gettext/manual/html_node/Locale-Environment-Variables.html).
I have a tentative fix for this that I am still testing.
